src/porepy/numerics/fv/mpfa.py

```diff
@@ -72,461 +69,260 @@ def discretize(self, g, data):
-        # outside the active faces. Prepare for this.
-        outside = np.setdiff1d(np.arange(g.num_faces), active_faces, assume_unique=True)
-
-        # Discretization of sub-problem
-        if not vector_source:
-            flux_loc, bound_flux_loc, bound_pressure_cell, bound_pressure_face = self._local_discr(
-                sub_g,
-                loc_k,
-                loc_bnd,
-                deviation_from_plane_tol,
-                vector_source=vector_source,
-                eta=eta,
-                eta_reconstruction=eta_reconstruction,
-                inverter=inverter,
-            )
-
-        else:
-            flux_loc, bound_flux_loc, bound_pressure_cell, bound_pressure_face, div_vector_source_loc = self._local_discr(
-                sub_g,
-                loc_k,
-                loc_bnd,
-                deviation_from_plane_tol,
-                vector_source=vector_source,
-                eta=eta,
-                eta_reconstruction=eta_reconstruction,
-                inverter=inverter,
+        # Update global face fields.
+        flux_glob = face_map * active_flux * cell_map
+        bound_flux_glob = face_map * active_bound_flux * face_map.transpose()
+        bound_pressure_cell_glob = face_map * active_bound_pressure_cell * cell_map
+        bound_pressure_face_glob = (
+            face_map * active_bound_pressure_face * face_map.transpose()
+        )
+        # The vector source term has a vector-sized number of rows, and need a
+        # different cell_map
+        _, cell_map_vec = pp.fvutils.map_subgrid_to_grid(
+            g, extracted_faces, active_cells, is_vector=True, nd=max(g.dim, 1)
+        )
+        vector_source_glob = face_map * active_vector_source * cell_map_vec
+
+        # The vector source term was computed in the natural coordinates of the grid.
+        # If the grid is embedded in a higher dimension, the source term will have the
+        # size of that dimensions, and the discretization matrix should be adjusted
+        # accordingly. The dimension of the vector source term is given by the parameter
+        # vector_source_dim.
+        # If the dimension of the grid is 3 (i.e. the maximum possible) or 0 (no vectors
+        # are possible), there is no need for adjustment.
+        if g.dim == 2:
+            # Use the same mapping of the geometry as was done in
+            # self._flux_discretization(). This mapping is deterministic, thus the
+            # rotation matrix should be the same as applied before. In this case, we
+            # only need the rotation, and the active dimensions
+            *_, R, dim, _ = pp.map_geometry.map_grid(g)
+
+            # We need to pick out the parts of the rotation matrix that gives the
+            # in-plane (relative to the grid) parts, and apply this to all cells in the
+            # grid. The simplest way to do this is to expand R[dim] via sps.block_diags,
```

Comment:
Where is the normal part handled?

Reply:
Not sure - perhaps it enters the interface law? Something for you, I am afraid.